### PR TITLE
Don't revalidate original rcParams when exiting rc_context.

### PR DESCRIPTION
**Pull request by anntzer**
- Rewrite rc_context using contextmanager, which removes the need for
  saving state.

- When `__exit__`ing rc_context, skip rcParam validation, as the items
  have already been validated originally.

The rationale for not revalidating (beyond a minor gain in speed) is
that it may make sense to smuggle seemingly "invalid" values into the
rcParams (by calling `dict.__setitem__`), and we do not want them to
cause an error when exiting a rc_context (of course, they may or may not
cause an error at the backend level, but that is the responsibility of
whoever set this "invalid" value).

For example, the rcparam `lines.antialiased` is declared as a boolean,
but the mpl_cairo backend actually also recognizes any of cairo's
internal antialiasing enum values (NONE, FAST, GOOD, BEST, etc.).

    
(The rst docs needed to be slightly updated because .. autoclass::
rc_context won't work anymore, so I also took advantage of that to
reorder all the rc-related functions slightly more logically.)

## PR Checklist

- [ ] Has Pytest style unit tests
- [X] Code is PEP 8 compliant 
- [X] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

File: doc/api/matplotlib_configuration_api.rst [hunk 1]
```diff
@@ -12,26 +12,24 @@ The top level :mod:`matplotlib` module
 
     An instance of :class:`RcParams` for handling default matplotlib values.
 
-.. autofunction:: rc
-
-.. autofunction::rcdefaults
-
-.. autofunction::rc_file
+.. autodata:: rc_context
```
Comment (tacaswell) [inline doc/api/matplotlib_configuration_api.rst]: autodata?

Comment (anntzer) [inline doc/api/matplotlib_configuration_api.rst]: See http://www.sphinx-doc.org/en/stable/ext/autodoc.html?highlight=autodata#directive-autofunction, but I guess this technically still qualifies as a function.

**Comment by anntzer** [comment]
Looks like s-g fails to resolve https://docs.{numpy,scipy}.org/doc/numpy/reference/searchindex.js which indeed do not exist.  Not sure whether they used to exist and numpy/scipy changed their docs build?  (the issue also appears with s-g 0.1.11 so it's not due to the s-g update)

**Comment by QuLogic** [comment]
You'll need to rebase against #8955 (or just latest `master`).

**Comment by anntzer** [comment]
thanks, done

**Pull request merged**